File: frontend/src/lib/api/babel.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface BabelInterface {
  name: string;
  type?: string | null; // auto, wired, wireless
  channel?: string | null; // 1-254, interfering, non-interfering
  hello_interval?: number | null; // 20-655340 ms
  update_interval?: number | null; // 20-655340 ms
  rxcost?: number | null; // 1-65534
  split_horizon?: string | null; // default, enable, disable
  enable_timestamps: boolean;
  max_rtt_penalty?: number | null; // 0-65535 ms
  rtt_decay?: number | null; // 1-256
  rtt_min?: number | null; // 1-65535 ms
  rtt_max?: number | null; // 1-65535 ms
}

export interface BabelParameters {
  diversity: boolean;
  diversity_factor?: number | null; // 1-256
  resend_delay?: number | null; // 20-655340 ms
  smoothing_half_life?: number | null; // 0-65534 seconds
}

export interface BabelRedistribute {
  ipv4: string[];
  ipv6: string[];
}

export interface DistributeListFilter {
  access_list_in?: string | null;
  access_list_out?: string | null;
  prefix_list_in?: string | null;
  prefix_list_out?: string | null;
}

export interface DistributeListInterfaceFilter {
  interface: string;
  access_list_in?: string | null;
  access_list_out?: string | null;
  prefix_list_in?: string | null;
  prefix_list_out?: string | null;
}

export interface BabelDistributeList {
  ipv4: DistributeListFilter;
  ipv6: DistributeListFilter;
  ipv4_interfaces: DistributeListInterfaceFilter[];
  ipv6_interfaces: DistributeListInterfaceFilter[];
}

export interface BabelConfig {
  interfaces: BabelInterface[];
  parameters: BabelParameters;
  redistribute: BabelRedistribute;
  distribute_list: BabelDistributeList;
}

export interface BabelCapabilities {
  version: string;
  features: {
    interfaces: { supported: boolean; description: string };
    parameters: { supported: boolean; description: string };
    redistribute: { supported: boolean; description: string };
    distribute_list: { supported: boolean; description: string };
    redistribute_nhrp: { supported: boolean; description: string };
  };
  version_info: {
    is_1_4: boolean;
    is_1_5: boolean;
  };
  redistribute_protocols: {
    ipv4: string[];
    ipv6: string[];
  };
  instance_name?: string;
  instance_id?: string;
}

export interface BabelBatchOperation {
  op: string;
  value?: string;
}

export interface BabelBatchRequest {
  operations: BabelBatchOperation[];
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

// ============================================================================
// API Service
// ============================================================================

class BabelService {
  async getCapabilities(): Promise<BabelCapabilities> {
    return apiClient.get<BabelCapabilities>("/vyos/babel/capabilities");
  }

  async getConfig(refresh = false): Promise<BabelConfig> {
    return apiClient.get<BabelConfig>("/vyos/babel/config", {
      refresh: refresh.toString(),
    });
  }

  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post<VyOSResponse>("/vyos/config/refresh");
  }

  async batchConfigure(request: BabelBatchRequest): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/babel/batch", request);
    await this.refreshConfig();
    return result;
  }

  // ==========================================================================
  // Interface Operations
  // ==========================================================================

  async createInterface(config: BabelInterface): Promise<VyOSResponse> {
    const operations: BabelBatchOperation[] = [
      { op: "set_interface", value: config.name },
    ];

    if (config.type) {
      operations.push({ op: "set_interface_type", value: `${config.name},${config.type}` });
    }
    if (config.channel) {
      operations.push({ op: "set_interface_channel", value: `${config.name},${config.channel}` });
    }
    if (config.hello_interval) {
      operations.push({ op: "set_interface_hello_interval", value: `${config.name},${config.hello_interval}` });
    }
    if (config.update_interval) {
      operations.push({ op: "set_interface_update_interval", value: `${config.name},${config.update_interval}` });
    }
    if (config.rxcost) {
      operations.push({ op: "set_interface_rxcost", value: `${config.name},${config.rxcost}` });
    }
    if (config.split_horizon) {
      operations.push({ op: "set_interface_split_horizon", value: `${config.name},${config.split_horizon}` });
    }
    if (config.enable_timestamps) {
      operations.push({ op: "set_interface_enable_timestamps", value: config.name });
    }
    if (config.max_rtt_penalty != null) {
      operations.push({ op: "set_interface_max_rtt_penalty", value: `${config.name},${config.max_rtt_penalty}` });
    }
    if (config.rtt_decay) {
      operations.push({ op: "set_interface_rtt_decay", value: `${config.name},${config.rtt_decay}` });
    }
    if (config.rtt_min) {
      operations.push({ op: "set_interface_rtt_min", value: `${config.name},${config.rtt_min}` });
    }
    if (config.rtt_max) {
      operations.push({ op: "set_interface_rtt_max", value: `${config.name},${config.rtt_max}` });
    }

    return this.batchConfigure({ operations });
  }

  async updateInterface(
    original: BabelInterface,
    updated: BabelInterface
  ): Promise<VyOSResponse> {
    const operations: BabelBatchOperation[] = [];
    const name = original.name;

    // Type
    if (updated.type !== original.type) {
      if (updated.type) {
        operations.push({ op: "set_interface_type", value: `${name},${updated.type}` });
      } else {
        operations.push({ op: "delete_interface_type", value: name });
      }
    }

    // Channel
    if (updated.channel !== original.channel) {
      if (updated.channel) {
        operations.push({ op: "set_interface_channel", value: `${name},${updated.channel}` });
      } else {
        operations.push({ op: "delete_interface_channel", value: name });
      }
    }

    // Hello interval
    if (updated.hello_interval !== original.hello_interval) {
      if (updated.hello_interval) {
        operations.push({ op: "set_interface_hello_interval", value: `${name},${updated.hello_interval}` });
      } else {
        operations.push({ op: "delete_interface_hello_interval", value: name });
      }
    }

    // Update interval
    if (updated.update_interval !== original.update_interval) {
      if (updated.update_interval) {
        operations.push({ op: "set_interface_update_interval", value: `${name},${updated.update_interval}` });
      } else {
        operations.push({ op: "delete_interface_update_interval", value: name });
      }
    }

    // Rxcost
    if (updated.rxcost !== original.rxcost) {
      if (updated.rxcost) {
        operations.push({ op: "set_interface_rxcost", value: `${name},${updated.rxcost}` });
      } else {
        operations.push({ op: "delete_interface_rxcost", value: name });
      }
    }

    // Split horizon
    if (updated.split_horizon !== original.split_horizon) {
      if (updated.split_horizon) {
        operations.push({ op: "set_interface_split_horizon", value: `${name},${updated.split_horizon}` });
      } else {
        operations.push({ op: "delete_interface_split_horizon", value: name });
      }
    }

    // Enable timestamps
    if (updated.enable_timestamps !== original.enable_timestamps) {
      if (updated.enable_timestamps) {
        operations.push({ op: "set_interface_enable_timestamps", value: name });
      } else {
        operations.push({ op: "delete_interface_enable_timestamps", value: name });
      }
    }

    // Max RTT penalty
    if (updated.max_rtt_penalty !== original.max_rtt_penalty) {
      if (updated.max_rtt_penalty != null) {
        operations.push({ op: "set_interface_max_rtt_penalty", value: `${name},${updated.max_rtt_penalty}` });
      } else {
        operations.push({ op: "delete_interface_max_rtt_penalty", value: name });
      }
    }

    // RTT decay
    if (updated.rtt_decay !== original.rtt_decay) {
      if (updated.rtt_decay) {
        operations.push({ op: "set_interface_rtt_decay", value: `${name},${updated.rtt_decay}` });
      } else {
        operations.push({ op: "delete_interface_rtt_decay", value: name });
      }
    }

    // RTT min
    if (updated.rtt_min !== original.rtt_min) {
      if (updated.rtt_min) {
        operations.push({ op: "set_interface_rtt_min", value: `${name},${updated.rtt_min}` });
      } else {
        operations.push({ op: "delete_interface_rtt_min", value: name });
      }
    }

    // RTT max
    if (updated.rtt_max !== original.rtt_max) {
      if (updated.rtt_max) {
        operations.push({ op: "set_interface_rtt_max", value: `${name},${updated.rtt_max}` });
      } else {
        operations.push({ op: "delete_interface_rtt_max", value: name });
      }
    }

    if (operations.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }

    return this.batchConfigure({ operations });
  }

  async deleteInterface(name: string): Promise<VyOSResponse> {
    return this.batchConfigure({
      operations: [{ op: "delete_interface", value: name }],
    });
  }

  // ==========================================================================
  // Parameters Operations
  // ==========================================================================

  async updateParameters(
    original: BabelParameters,
    updated: BabelParameters
  ): Promise<VyOSResponse> {
    const operations: BabelBatchOperation[] = [];

    if (updated.diversity !== original.diversity) {
      operations.push({
        op: updated.diversity ? "set_parameters_diversity" : "delete_parameters_diversity",
      });
    }

    if (updated.diversity_factor !== original.diversity_factor) {
      if (updated.diversity_factor) {
        operations.push({ op: "set_parameters_diversity_factor", value: String(updated.diversity_factor) });
      } else {
        operations.push({ op: "delete_parameters_diversity_factor" });
      }
    }

    if (updated.resend_delay !== original.resend_delay) {
      if (updated.resend_delay) {
        operations.push({ op: "set_parameters_resend_delay", value: String(updated.resend_delay) });
      } else {
        operations.push({ op: "delete_parameters_resend_delay" });
      }
    }

    if (updated.smoothing_half_life !== original.smoothing_half_life) {
      if (updated.smoothing_half_life != null) {
        operations.push({ op: "set_parameters_smoothing_half_life", value: String(updated.smoothing_half_life) });
      } else {
        operations.push({ op: "delete_parameters_smoothing_half_life" });
      }
    }

    if (operations.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }

    return this.batchConfigure({ operations });
  }

  // ==========================================================================
  // Redistribute Operations
  // ==========================================================================

  async updateRedistribute(
    original: BabelRedistribute,
    updated: BabelRedistribute
  ): Promise<VyOSResponse> {
    const operations: BabelBatchOperation[] = [];

    // IPv4 removals
    for (const proto of original.ipv4) {
      if (!updated.ipv4.includes(proto)) {
        operations.push({ op: "delete_redistribute_ipv4", value: proto });
      }
    }
    // IPv4 additions
    for (const proto of updated.ipv4) {
      if (!original.ipv4.includes(proto)) {
        operations.push({ op: "set_redistribute_ipv4", value: proto });
      }
    }

    // IPv6 removals
    for (const proto of original.ipv6) {
      if (!updated.ipv6.includes(proto)) {
        operations.push({ op: "delete_redistribute_ipv6", value: proto });
      }
    }
    // IPv6 additions
    for (const proto of updated.ipv6) {
      if (!original.ipv6.includes(proto)) {
        operations.push({ op: "set_redistribute_ipv6", value: proto });
      }
    }

    if (operations.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }

    return this.batchConfigure({ operations });
  }

  // ==========================================================================
  // Distribute List Operations
  // ==========================================================================

  async updateDistributeListGlobal(
    af: "ipv4" | "ipv6",
    original: DistributeListFilter,
    updated: DistributeListFilter
  ): Promise<VyOSResponse> {
    const operations: BabelBatchOperation[] = [];
    const prefix = `distribute_list_${af}`;

    if (updated.access_list_in !== original.access_list_in) {
      if (updated.access_list_in) {
        operations.push({ op: `set_${prefix}_access_list_in`, value: updated.access_list_in });
      } else {
        operations.push({ op: `delete_${prefix}_access_list_in` });
      }
    }

    if (updated.access_list_out !== original.access_list_out) {
      if (updated.access_list_out) {
        operations.push({ op: `set_${prefix}_access_list_out`, value: updated.access_list_out });
      } else {
        operations.push({ op: `delete_${prefix}_access_list_out` });
      }
    }

    if (updated.prefix_list_in !== original.prefix_list_in) {
      if (updated.prefix_list_in) {
        operations.push({ op: `set_${prefix}_prefix_list_in`, value: updated.prefix_list_in });
      } else {
        operations.push({ op: `delete_${prefix}_prefix_list_in` });
      }
    }

    if (updated.prefix_list_out !== original.prefix_list_out) {
      if (updated.prefix_list_out) {
        operations.push({ op: `set_${prefix}_prefix_list_out`, value: updated.prefix_list_out });
      } else {
        operations.push({ op: `delete_${prefix}_prefix_list_out` });
      }
    }

    if (operations.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }

    return this.batchConfigure({ operations });
  }

  async updateDistributeListInterface(
    af: "ipv4" | "ipv6",
    interfaceName: string,
    original: DistributeListInterfaceFilter | null,
    updated: DistributeListInterfaceFilter
  ): Promise<VyOSResponse> {
    const operations: BabelBatchOperation[] = [];
    const prefix = `distribute_list_${af}_iface`;
    const orig = original || { interface: interfaceName, access_list_in: null, access_list_out: null, prefix_list_in: null, prefix_list_out: null };

    if (updated.access_list_in !== orig.access_list_in) {
      if (updated.access_list_in) {
        operations.push({ op: `set_${prefix}_access_list_in`, value: `${interfaceName},${updated.access_list_in}` });
      } else {
        operations.push({ op: `delete_${prefix}_access_list_in`, value: interfaceName });
      }
    }

    if (updated.access_list_out !== orig.access_list_out) {
      if (updated.access_list_out) {
        operations.push({ op: `set_${prefix}_access_list_out`, value: `${interfaceName},${updated.access_list_out}` });
      } else {
        operations.push({ op: `delete_${prefix}_access_list_out`, value: interfaceName });
      }
    }

    if (updated.prefix_list_in !== orig.prefix_list_in) {
      if (updated.prefix_list_in) {
        operations.push({ op: `set_${prefix}_prefix_list_in`, value: `${interfaceName},${updated.prefix_list_in}` });
      } else {
        operations.push({ op: `delete_${prefix}_prefix_list_in`, value: interfaceName });
      }
    }

    if (updated.prefix_list_out !== orig.prefix_list_out) {
      if (updated.prefix_list_out) {
        operations.push({ op: `set_${prefix}_prefix_list_out`, value: `${interfaceName},${updated.prefix_list_out}` });
      } else {
        operations.push({ op: `delete_${prefix}_prefix_list_out`, value: interfaceName });
      }
    }

    if (operations.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }

    return this.batchConfigure({ operations });
  }
}

export const babelService = new BabelService();

File: backend/routers/babel/babel.py
"""
Babel Protocol Router

API endpoints for managing VyOS Babel routing protocol configuration.
Supports version-aware configuration for VyOS 1.4 and 1.5.
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import BabelBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
import inspect

router = APIRouter(prefix="/vyos/babel", tags=["babel"])

# Builder infrastructure methods that must never be invokable via the batch API
_INTERNAL_BUILDER_METHODS = frozenset({
    "add_set", "add_delete", "get_operations", "is_empty", "clear", "operation_count",
})


# ============================================================================
# Pydantic Models
# ============================================================================


class BabelInterface(BaseModel):
    """Babel interface configuration."""
    name: str
    type: Optional[str] = None  # auto, wired, wireless
    channel: Optional[str] = None  # 1-254, interfering, non-interfering
    hello_interval: Optional[int] = None  # 20-655340 ms
    update_interval: Optional[int] = None  # 20-655340 ms
    rxcost: Optional[int] = None  # 1-65534
    split_horizon: Optional[str] = None  # default, enable, disable
    enable_timestamps: bool = False
    max_rtt_penalty: Optional[int] = None  # 0-65535 ms
    rtt_decay: Optional[int] = None  # 1-256
    rtt_min: Optional[int] = None  # 1-65535 ms
    rtt_max: Optional[int] = None  # 1-65535 ms


class BabelParameters(BaseModel):
    """Babel global parameters."""
    diversity: bool = False
    diversity_factor: Optional[int] = None  # 1-256
    resend_delay: Optional[int] = None  # 20-655340 ms
    smoothing_half_life: Optional[int] = None  # 0-65534 seconds


class BabelRedistribute(BaseModel):
    """Babel redistribution settings."""
    ipv4: List[str] = []  # List of protocols
    ipv6: List[str] = []  # List of protocols


class DistributeListFilter(BaseModel):
    """A single direction filter (access-list or prefix-list)."""
    access_list_in: Optional[str] = None
    access_list_out: Optional[str] = None
    prefix_list_in: Optional[str] = None
    prefix_list_out: Optional[str] = None


class DistributeListInterfaceFilter(BaseModel):
    """Per-interface distribute list filter."""
    interface: str
    access_list_in: Optional[str] = None
    access_list_out: Optional[str] = None
    prefix_list_in: Optional[str] = None
    prefix_list_out: Optional[str] = None


class BabelDistributeList(BaseModel):
    """Babel distribute list configuration."""
    ipv4: DistributeListFilter = DistributeListFilter()
    ipv6: DistributeListFilter = DistributeListFilter()
    ipv4_interfaces: List[DistributeListInterfaceFilter] = []
    ipv6_interfaces: List[DistributeListInterfaceFilter] = []


class BabelConfig(BaseModel):
    """Complete Babel configuration."""
    interfaces: List[BabelInterface] = []
    parameters: BabelParameters = BabelParameters()
    redistribute: BabelRedistribute = BabelRedistribute()
    distribute_list: BabelDistributeList = BabelDistributeList()


class BabelBatchOperation(BaseModel):
    """Single operation in a batch request."""
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value")


class BabelBatchRequest(BaseModel):
    """Model for batch configuration."""
    operations: List[BabelBatchOperation]


class VyOSResponse(BaseModel):
    """Standard response from VyOS operations."""
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_babel_capabilities(request: Request):
    """Get Babel feature capabilities based on device VyOS version."""
    await require_read_permission(request, FeatureGroup.BABEL)

    try:
        service = get_session_vyos_service(request)
        version = service.get_version()
        builder = BabelBatchBuilder(version=version)
        capabilities = builder.get_capabilities()

        if hasattr(request.state, "instance") and request.state.instance:
            capabilities["instance_name"] = request.state.instance.get("name")
            capabilities["instance_id"] = request.state.instance.get("id")
        return capabilities
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Endpoint 2: Config
# ============================================================================


@router.get("/config", response_model=BabelConfig)
async def get_babel_config(http_request: Request, refresh: bool = False):
    """Get all Babel configuration from VyOS in a generalized format."""
    await require_read_permission(http_request, FeatureGroup.BABEL)

    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        babel_config = full_config.get("protocols", {}).get("babel", {})

        if not babel_config:
            return BabelConfig()

        # Parse interfaces
        interfaces = parse_interfaces(babel_config.get("interface", {}))

        # Parse parameters
        parameters = parse_parameters(babel_config.get("parameters", {}))

        # Parse redistribute
        redistribute = parse_redistribute(babel_config.get("redistribute", {}))

        # Parse distribute-list
        distribute_list = parse_distribute_list(babel_config.get("distribute-list", {}))

        return BabelConfig(
            interfaces=interfaces,
            parameters=parameters,
            redistribute=redistribute,
            distribute_list=distribute_list
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Config Parsers
# ============================================================================


def parse_interfaces(interfaces_raw: dict) -> List[BabelInterface]:
    """Parse Babel interface configuration."""
    interfaces = []

    for iface_name, iface_config in interfaces_raw.items():
        if iface_config is None:
            iface_config = {}

        interfaces.append(BabelInterface(
            name=iface_name,
            type=iface_config.get("type"),
            channel=iface_config.get("channel"),
            hello_interval=int(iface_config["hello-interval"]) if iface_config.get("hello-interval") else None,
            update_interval=int(iface_config["update-interval"]) if iface_config.get("update-interval") else None,
            rxcost=int(iface_config["rxcost"]) if iface_config.get("rxcost") else None,
            split_horizon=iface_config.get("split-horizon"),
            enable_timestamps="enable-timestamps" in iface_config,
            max_rtt_penalty=int(iface_config["max-rtt-penalty"]) if iface_config.get("max-rtt-penalty") else None,
            rtt_decay=int(iface_config["rtt-decay"]) if iface_config.get("rtt-decay") else None,
            rtt_min=int(iface_config["rtt-min"]) if iface_config.get("rtt-min") else None,
            rtt_max=int(iface_config["rtt-max"]) if iface_config.get("rtt-max") else None,
        ))

    return interfaces


def parse_parameters(params_raw: dict) -> BabelParameters:
    """Parse Babel global parameters."""
    if not params_raw:
        return BabelParameters()

    return BabelParameters(
        diversity="diversity" in params_raw,
        diversity_factor=int(params_raw["diversity-factor"]) if params_raw.get("diversity-factor") else None,
        resend_delay=int(params_raw["resend-delay"]) if params_raw.get("resend-delay") else None,
        smoothing_half_life=int(params_raw["smoothing-half-life"]) if params_raw.get("smoothing-half-life") else None,
    )


def parse_redistribute(redistribute_raw: dict) -> BabelRedistribute:
    """Parse Babel redistribution settings."""
    if not redistribute_raw:
        return BabelRedistribute()

    ipv4_protocols = list(redistribute_raw.get("ipv4", {}).keys())
    ipv6_protocols = list(redistribute_raw.get("ipv6", {}).keys())

    return BabelRedistribute(
        ipv4=ipv4_protocols,
        ipv6=ipv6_protocols
    )


def parse_distribute_list(dist_list_raw: dict) -> BabelDistributeList:
    """Parse Babel distribute list configuration."""
    if not dist_list_raw:
        return BabelDistributeList()

    ipv4_raw = dist_list_raw.get("ipv4", {})
    ipv6_raw = dist_list_raw.get("ipv6", {})

    # Parse global filters
    ipv4_filter = _parse_global_filter(ipv4_raw)
    ipv6_filter = _parse_global_filter(ipv6_raw)

    # Parse per-interface filters
    ipv4_iface_filters = _parse_interface_filters(ipv4_raw.get("interface", {}))
    ipv6_iface_filters = _parse_interface_filters(ipv6_raw.get("interface", {}))

    return BabelDistributeList(
        ipv4=ipv4_filter,
        ipv6=ipv6_filter,
        ipv4_interfaces=ipv4_iface_filters,
        ipv6_interfaces=ipv6_iface_filters
    )


def _parse_global_filter(af_raw: dict) -> DistributeListFilter:
    """Parse global (non-interface) distribute-list filter."""
    if not af_raw:
        return DistributeListFilter()

    acl_raw = af_raw.get("access-list", {})
    pfx_raw = af_raw.get("prefix-list", {})

    return DistributeListFilter(
        access_list_in=acl_raw.get("in") if isinstance(acl_raw, dict) else None,
        access_list_out=acl_raw.get("out") if isinstance(acl_raw, dict) else None,
        prefix_list_in=pfx_raw.get("in") if isinstance(pfx_raw, dict) else None,
        prefix_list_out=pfx_raw.get("out") if isinstance(pfx_raw, dict) else None,
    )


def _parse_interface_filters(iface_raw: dict) -> List[DistributeListInterfaceFilter]:
    """Parse per-interface distribute-list filters."""
    filters = []

    if not iface_raw:
        return filters

    for iface_name, iface_config in iface_raw.items():
        if iface_config is None:
            iface_config = {}

        acl_raw = iface_config.get("access-list", {})
        pfx_raw = iface_config.get("prefix-list", {})

        if not acl_raw:
            acl_raw = {}
        if not pfx_raw:
            pfx_raw = {}

        filters.append(DistributeListInterfaceFilter(
            interface=iface_name,
            access_list_in=acl_raw.get("in") if isinstance(acl_raw, dict) else None,
            access_list_out=acl_raw.get("out") if isinstance(acl_raw, dict) else None,
            prefix_list_in=pfx_raw.get("in") if isinstance(pfx_raw, dict) else None,
            prefix_list_out=pfx_raw.get("out") if isinstance(pfx_raw, dict) else None,
        ))

    return filters


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch", response_model=VyOSResponse)
async def babel_batch_configure(http_request: Request, body: BabelBatchRequest):
    """Execute a batch of Babel configuration operations."""
    await require_write_permission(http_request, FeatureGroup.BABEL)

    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = BabelBatchBuilder(version=version)

        for operation in body.operations:
            if operation.op.startswith("_") or operation.op in _INTERNAL_BUILDER_METHODS:
                raise HTTPException(status_code=400, detail=f"Invalid operation: {operation.op}")
            method = getattr(builder, operation.op, None)
            if not callable(method):
                raise HTTPException(status_code=400, detail=f"Unknown operation: {operation.op}")
            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 0:
                method()
            elif len(params) == 1 and operation.value:
                method(operation.value)
            elif len(params) == 2 and operation.value:
                values = operation.value.split(",", 1)
                if len(values) == 2:
                    method(values[0], values[1])
                else:
                    method(operation.value)
            elif len(params) == 0:
                method()

        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Babel configuration updated"},
            error=response.error if response.error else None
        )
    except AttributeError as e:
        raise HTTPException(status_code=400, detail=f"Unknown operation: {str(e)}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
